tasks/framework-tools/src/utils.js

```diff
@@ -0,0 +1,238 @@
+      )) {
+        // Skip `@redwoodjs/*` packages, since these are processed
+        // by the workspace.
+        if (!name.startsWith('@redwoodjs/')) {
+          if (dependencies[name] && dependencies[name] !== version) {
+            warnings.push([
+              name,
+              'dependency version mismatched, please make sure the versions are the same.',
+            ])
+          }
+          dependencies[name] = version
+        }
+      }
+    } catch (error) {
+      console.error()
+      console.error(c.red(`Error in ${packageFile}:`))
+      console.error('', error.message)
+      console.error()
+      console.error(` This package's dependencies will not be included.`)
+      console.error()
+    }
+  }
+  return { dependencies: sortObjectKeys(dependencies), warnings }
+}
+
+function logWarnings(warnings) {
+  for (const [packageName, message] of warnings) {
+    console.warn(c.yellow('Warning:'), packageName, message)
+  }
+}
+
+/**
+ * The files that are published by the Redwood Framework.
+ * Note: The packages must be built.
+ */
+function packagesFileList(packages = frameworkPackages()) {
+  const fileList = {}
+  for (const packageFile of packages) {
+    const packageJson = require(packageFile)
+
+    if (!packageJson.name) {
+      continue
+    }
+
+    fileList[packageJson.name] = packlist.sync({
+      path: path.dirname(packageFile),
+    })
+  }
+  return fileList
+}
+
+function redwoodBins(packages = frameworkPackages()) {
+  let bins = {}
+  for (const packageFile of packages) {
+    const packageJson = require(packageFile)
+    if (!packageJson.name) {
+      continue
+    }
+
+    if (!packageJson.bin) {
+      continue
+    }
+    for (const [binName, binPath] of Object.entries(packageJson.bin)) {
+      bins[binName] = path.join(packageJson.name, binPath)
+    }
+  }
+  return bins
+}
+
+// gets the redwood project's package.json and builds a write function around it
+function getPackageJson(projectPath) {
+  const packageJsonPath = path.join(projectPath, 'package.json')
+
+  const packageJsonLink = terminalLink(
+    'Redwood Project Path',
+    'file://' + packageJsonPath
+  )
+
+  if (!fs.existsSync(packageJsonPath)) {
+    console.log(
+      `Error: The specified ${packageJsonLink} does not have a package.json file.`
+    )
+    console.log('Expected:', packageJsonPath)
+    process.exit(1)
+  }
+
+  const packageJson = require(packageJsonPath)
+
+  // for use later, after updating packageJson's dependencies
+  function writePackageJson(packageJson) {
+    fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, undefined, 2))
+  }
+
+  function updatePackageJsonDeps(dependencies) {
+    return produce(packageJson, (draftPackageJson) => {
+      if (draftPackageJson.dependencies) {
+        draftPackageJson.dependencies = {
+          ...draftPackageJson.dependencies,
+          ...dependencies,
+        }
+      } else {
+        draftPackageJson.dependencies = dependencies
+      }
+    })
```

Comment:
I understand that you might want to use immer when you want to produce the next state tree in memory, but I can't understand the usage here?

Reply:
On exit, we should restore the user's original package.json. But during sync I want to make immutable updates to the package.json based on new dependencies installed. Immer just lets me do that while keeping the old one for restoring on exit.